instant-segment/src/lib.rs

```diff
@@ -13,47 +13,55 @@ pub mod test_data;
 /// Central data structure used to calculate word probabilities
 #[cfg_attr(feature = "with-serde", derive(Deserialize, Serialize))]
 pub struct Segmenter {
-    unigrams: HashMap<String, f64>,
-    bigrams: HashMap<(String, String), f64>,
+    // Maps a word to both its unigram score, as well has a nested HashMap in
+    // which the bigram score can be looked up using the previous word. Scores
+    // are base-10 logarithms of relative word frequencies
+    scores: HashMap<String, (f64, HashMap<String, f64>)>,
+    // Base-10 logarithm of the total count of unigrams
     uni_total_log10: f64,
     limit: usize,
 }
 
 impl Segmenter {
-    /// Create `Segmenter` from the given iterators
+    /// Create `Segmenter` from the given unigram and bigram counts.
     ///
     /// Note: the `String` types used in this API are defined in the `smartstring` crate. Any
     /// `&str` or `String` can be converted into the `String` used here by calling `into()` on it.
-    pub fn from_iters<U, B>(unigrams: U, bigrams: B) -> Self
+    pub fn new<U, B>(unigrams: U, bigrams: B) -> Self
     where
-        U: Iterator<Item = (String, f64)>,
-        B: Iterator<Item = ((String, String), f64)>,
+        U: IntoIterator<Item = (String, f64)>,
+        B: IntoIterator<Item = ((String, String), f64)>,
     {
-        Self::from_maps(unigrams.collect(), bigrams.collect())
-    }
+        // Initially, `scores` contains the original unigram and bigram counts
+        let mut scores = unigrams
+            .into_iter()
+            .map(|(word, uni)| (word, (uni, HashMap::default())))
+            .collect::<HashMap<_, _>>();
+        let uni_total: f64 = scores.values().map(|(score, _)| *score).sum();
+        let mut bi_total = 0.0;
+        for ((word1, word2), bi) in bigrams {
+            if let Some((_, bi_scores)) = scores.get_mut(&word2) {
```

Comment:
Style nit: let's write this with an early continue -- we can probably use `let .. else` for this. Move the comment about throwing things away into the block with the `continue`.

Reply:
I wasn't sure what the minimum supported Rust version of this crate is - `let ... else` was stabilized in Rust 1.65 (released Nov. 3, 2022) which is kinda new.
However, I agree that `let ... else` is cleaner, so I changed this and also added a separate commit before this one where I specify the MSRV in `Cargo.toml` of both crates to be 1.65. Let me know if you don't like this, then I can delete that commit.